File: Sources/HaviKit/HaviTokenStore.swift
import Foundation
#if canImport(Security)
import Security
#endif

/// Backing for `HaviTokenStore`'s string items. Production resolves the Keychain
/// backing; the poll-state-machine and persistence tests inject an in-memory
/// backing so nothing touches the real Keychain (design §1, §5).
protocol HaviCredentialBacking: Sendable {
    func read(_ account: String) -> String?
    func write(_ value: String, account: String)
    func delete(_ account: String)
}

/// The connected HAVI identity resolved by the device-code flow (design §5): the
/// bearer credential plus the display names the connect sheet shows in its
/// success + "Connected as …" states.
public struct HaviConnectedSession: Sendable, Equatable {
    public let accessToken: String
    public let workspaceID: String
    public let refreshToken: String?
    public let expiresAt: Date?
    public let userName: String?
    public let workspaceName: String?

    public init(
        accessToken: String,
        workspaceID: String,
        refreshToken: String? = nil,
        expiresAt: Date? = nil,
        userName: String? = nil,
        workspaceName: String? = nil
    ) {
        self.accessToken = accessToken
        self.workspaceID = workspaceID
        self.refreshToken = refreshToken
        self.expiresAt = expiresAt
        self.userName = userName
        self.workspaceName = workspaceName
    }
}

/// HaviKit's own credential store under a **distinct** service string
/// `ai.handgemacht.havikit`, kept separate from the app's `KeychainTokenStorage`
/// (`ai.handgemacht.lesewerkstatt`) so the dogfood identity never mixes with the
/// child's account and the two lifecycles are independent (design §1). The
/// stamped `devToken` is a fallback source only and is never written here.
/// Manual-paste and device-code credentials, when present, override the stamped
/// token.
public final class HaviTokenStore: @unchecked Sendable {
    private enum Key {
        static let accessToken = "havi.access_token"
        static let workspaceID = "havi.workspace_id"
        static let refreshToken = "havi.refresh_token"
        static let expiresAt = "havi.expires_at"
        static let userName = "havi.user_name"
        static let workspaceName = "havi.workspace_name"
    }

    private let backing: HaviCredentialBacking

    public init() {
        #if canImport(Security)
        self.backing = HaviKeychainBacking(service: "ai.handgemacht.havikit")
        #else
        self.backing = HaviInMemoryCredentialBacking()
        #endif
    }

    init(backing: HaviCredentialBacking) {
        self.backing = backing
    }

    public var accessToken: String? { backing.read(Key.accessToken) }
    public var workspaceID: String? { backing.read(Key.workspaceID) }
    public var refreshToken: String? { backing.read(Key.refreshToken) }
    public var userName: String? { backing.read(Key.userName) }
    public var workspaceName: String? { backing.read(Key.workspaceName) }

    public var expiresAt: Date? {
        guard let raw = backing.read(Key.expiresAt), let seconds = TimeInterval(raw) else { return nil }
        return Date(timeIntervalSince1970: seconds)
    }

    public var hasCredential: Bool {
        accessToken != nil && workspaceID != nil
    }

    /// The stored identity reassembled for the connect sheet's "Connected as …"
    /// row, or `nil` when no credential is present.
    public var connectedSession: HaviConnectedSession? {
        guard let accessToken, let workspaceID else { return nil }
        return HaviConnectedSession(
            accessToken: accessToken,
            workspaceID: workspaceID,
            refreshToken: refreshToken,
            expiresAt: expiresAt,
            userName: userName,
            workspaceName: workspaceName
        )
    }

    /// Manual-paste dev fallback (design §5): a bearer token + workspace id copied
    /// from the dashboard, overriding the stamped values.
    public func signIn(token: String, workspaceID: String) {
        backing.write(token, account: Key.accessToken)
        backing.write(workspaceID, account: Key.workspaceID)
        backing.delete(Key.refreshToken)
        backing.delete(Key.expiresAt)
        backing.delete(Key.userName)
        backing.delete(Key.workspaceName)
    }

    /// Device-code result (design §5): the full session, including the display
    /// names shown in the connect sheet.
    public func store(_ session: HaviConnectedSession) {
        backing.write(session.accessToken, account: Key.accessToken)
        backing.write(session.workspaceID, account: Key.workspaceID)
        writeOrDelete(session.refreshToken, account: Key.refreshToken)
        writeOrDelete(session.expiresAt.map { String($0.timeIntervalSince1970) }, account: Key.expiresAt)
        writeOrDelete(session.userName, account: Key.userName)
        writeOrDelete(session.workspaceName, account: Key.workspaceName)
    }

    public func clear() {
        backing.delete(Key.accessToken)
        backing.delete(Key.workspaceID)
        backing.delete(Key.refreshToken)
        backing.delete(Key.expiresAt)
        backing.delete(Key.userName)
        backing.delete(Key.workspaceName)
    }

    private func writeOrDelete(_ value: String?, account: String) {
        if let value {
            backing.write(value, account: account)
        } else {
            backing.delete(account)
        }
    }
}

#if canImport(Security)
/// Keychain-backed generic-password store used in production.
struct HaviKeychainBacking: HaviCredentialBacking {
    let service: String

    func read(_ account: String) -> String? {
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: service,
            kSecAttrAccount as String: account,
            kSecReturnData as String: true,
            kSecMatchLimit as String: kSecMatchLimitOne,
        ]
        var result: AnyObject?
        let status = SecItemCopyMatching(query as CFDictionary, &result)
        guard status == errSecSuccess, let data = result as? Data else { return nil }
        return String(data: data, encoding: .utf8)
    }

    /// Delete-then-add, so an item written before `kSecAttrAccessible` was set
    /// picks the attribute up on the next store. The credential is only ever used
    /// by a foreground capture on this device, so it needs neither background
    /// access before the first unlock nor a place in an iCloud/device backup.
    func write(_ value: String, account: String) {
        guard let data = value.data(using: .utf8) else { return }
        delete(account)
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: service,
            kSecAttrAccount as String: account,
            kSecValueData as String: data,
            kSecAttrAccessible as String: kSecAttrAccessibleAfterFirstUnlockThisDeviceOnly,
        ]
        SecItemAdd(query as CFDictionary, nil)
    }

    func delete(_ account: String) {
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: service,
            kSecAttrAccount as String: account,
        ]
        SecItemDelete(query as CFDictionary)
    }
}
#endif

/// Lock-guarded in-memory backing, injected by the tests (and the fallback on any
/// platform without `Security`).
final class HaviInMemoryCredentialBacking: HaviCredentialBacking, @unchecked Sendable {
    private let lock = NSLock()
    private var storage: [String: String] = [:]

    init() {}

    func read(_ account: String) -> String? {
        lock.lock(); defer { lock.unlock() }
        return storage[account]
    }

    func write(_ value: String, account: String) {
        lock.lock(); defer { lock.unlock() }
        storage[account] = value
    }

    func delete(_ account: String) {
        lock.lock(); defer { lock.unlock() }
        storage[account] = nil
    }
}

File: android/havikit-core/src/main/kotlin/ai/handgemacht/havikit/HaviTokenStore.kt
package ai.handgemacht.havikit

import java.time.Instant
import java.util.concurrent.locks.ReentrantLock
import kotlin.concurrent.withLock

/**
 * The connected HAVI identity resolved by the device-code flow (wire spec §4.3,
 * §11): the bearer credential plus the display names the connect sheet shows.
 * `refreshToken` is always null in practice — the exchange parser hard-codes null
 * and there is no refresh flow anywhere in the SDK.
 */
public data class HaviConnectedSession(
    val accessToken: String,
    val workspaceId: String,
    val refreshToken: String? = null,
    val expiresAt: Instant? = null,
    val userName: String? = null,
    val workspaceName: String? = null,
)

/**
 * Backing for the token store's string items. Production (Android) resolves an
 * EncryptedSharedPreferences / Keystore backing under a distinct namespace,
 * analogous to iOS's dedicated `ai.handgemacht.havikit` Keychain service; the core
 * ships an in-memory backing so the credential lifecycle is unit-tested without a
 * device.
 */
public interface HaviCredentialBacking {
    public fun read(account: String): String?

    public fun write(
        account: String,
        value: String,
    )

    public fun delete(account: String)
}

/**
 * HaviKit's own credential store, kept separate from the host app's store so the
 * dogfood identity never mixes with the child's account (wire spec §11). The
 * stamped `devToken` is a fallback source only and is never written here.
 * Manual-paste and device-code credentials, when present, override the stamped
 * token. `hasCredential` = access token AND workspace id both present.
 */
public class HaviTokenStore(
    private val backing: HaviCredentialBacking = InMemoryCredentialBacking(),
) {
    private object Key {
        const val ACCESS_TOKEN = "havi.access_token"
        const val WORKSPACE_ID = "havi.workspace_id"
        const val REFRESH_TOKEN = "havi.refresh_token"
        const val EXPIRES_AT = "havi.expires_at"
        const val USER_NAME = "havi.user_name"
        const val WORKSPACE_NAME = "havi.workspace_name"
    }

    public val accessToken: String? get() = backing.read(Key.ACCESS_TOKEN)
    public val workspaceId: String? get() = backing.read(Key.WORKSPACE_ID)
    public val refreshToken: String? get() = backing.read(Key.REFRESH_TOKEN)
    public val userName: String? get() = backing.read(Key.USER_NAME)
    public val workspaceName: String? get() = backing.read(Key.WORKSPACE_NAME)

    /** Stored as unix epoch seconds (wire spec §11); informational only. */
    public val expiresAt: Instant?
        get() = backing.read(Key.EXPIRES_AT)?.toDoubleOrNull()?.let { Instant.ofEpochMilli((it * 1000).toLong()) }

    public val hasCredential: Boolean get() = accessToken != null && workspaceId != null

    /** The stored identity reassembled for the connect sheet's "Connected as …" row. */
    public val connectedSession: HaviConnectedSession?
        get() {
            val token = accessToken ?: return null
            val workspace = workspaceId ?: return null
            return HaviConnectedSession(
                accessToken = token,
                workspaceId = workspace,
                refreshToken = refreshToken,
                expiresAt = expiresAt,
                userName = userName,
                workspaceName = workspaceName,
            )
        }

    /** Manual-paste dev fallback: a bearer token + workspace id, clearing the rest. */
    public fun signIn(
        token: String,
        workspaceId: String,
    ) {
        backing.write(Key.ACCESS_TOKEN, token)
        backing.write(Key.WORKSPACE_ID, workspaceId)
        backing.delete(Key.REFRESH_TOKEN)
        backing.delete(Key.EXPIRES_AT)
        backing.delete(Key.USER_NAME)
        backing.delete(Key.WORKSPACE_NAME)
    }

    /** Device-code result: the full session, writing present fields and deleting absent ones. */
    public fun store(session: HaviConnectedSession) {
        backing.write(Key.ACCESS_TOKEN, session.accessToken)
        backing.write(Key.WORKSPACE_ID, session.workspaceId)
        writeOrDelete(Key.REFRESH_TOKEN, session.refreshToken)
        writeOrDelete(Key.EXPIRES_AT, session.expiresAt?.let { epochSecondsString(it) })
        writeOrDelete(Key.USER_NAME, session.userName)
        writeOrDelete(Key.WORKSPACE_NAME, session.workspaceName)
    }

    public fun clear() {
        backing.delete(Key.ACCESS_TOKEN)
        backing.delete(Key.WORKSPACE_ID)
        backing.delete(Key.REFRESH_TOKEN)
        backing.delete(Key.EXPIRES_AT)
        backing.delete(Key.USER_NAME)
        backing.delete(Key.WORKSPACE_NAME)
    }

    private fun writeOrDelete(
        account: String,
        value: String?,
    ) {
        if (value != null) backing.write(account, value) else backing.delete(account)
    }

    private fun epochSecondsString(instant: Instant): String {
        val seconds = instant.toEpochMilli() / 1000.0
        return if (seconds == seconds.toLong().toDouble()) seconds.toLong().toString() else seconds.toString()
    }
}

/** Lock-guarded in-memory backing (the core default and the tests' backing). */
public class InMemoryCredentialBacking : HaviCredentialBacking {
    private val lock = ReentrantLock()
    private val storage = HashMap<String, String>()

    override fun read(account: String): String? = lock.withLock { storage[account] }

    override fun write(
        account: String,
        value: String,
    ): Unit = lock.withLock { storage[account] = value }

    override fun delete(account: String): Unit = lock.withLock { storage.remove(account) }
}
